internal/model/database.go

```diff
@@ -0,0 +1,281 @@
+	// uploaded and if so will set the is_uploaded flag to true
+	//
+	// Arguments:
+	//
+	// - result is the database result to update
+	//
+	// Returns a non-nil error if update failed
+	UpdateUploadedStatus(result *DatabaseResult) error
+
+	// Finished marks the result as done and sets the runtime
+	//
+	// Arguments:
+	//
+	// - result is the database result to mark as done
+	//
+	// Returns a non-nil error if result could not be marked as done
+	Finished(result *DatabaseResult) error
+
+	// DeleteResult will delete a particular result and the relative measurement on
+	// disk.
+	//
+	// Arguments:
+	//
+	// - resultID is the id of the database result to be deleted
+	//
+	// Returns a non-nil error if result could not be deleted
+	DeleteResult(resultID int64) error
+
+	// CreateMeasurement writes the measurement to the database a returns a pointer
+	// to the Measurement
+	//
+	// Arguments:
+	//
+	// - reportID is the report id to annotate
+	//
+	// - testName is the experiment name to use
+	//
+	// - measurementDir is the measurement directory path
+	//
+	// - resultID is the result id to annotate
+	//
+	// - urlID is the id of the URL input
+	//
+	// Returns either a database measurement or an error
+	CreateMeasurement(reportID sql.NullString, testName string, measurementDir string, idx int,
+		resultID int64, urlID sql.NullInt64) (*DatabaseMeasurement, error)
+
+	// AddTestKeys writes the summary to the measurement
+	//
+	// Arguments:
+	//
+	// - msmt is the database measurement to update
+	//
+	// - tk is the testkeys
+	//
+	// Returns a non-nil error if measurement update failed
+	AddTestKeys(msmt *DatabaseMeasurement, tk interface{}) error
+
+	// Done marks the measurement as completed
+	//
+	// Arguments:
+	//
+	// - msmt is the database measurement to update
+	//
+	// Returns a non-nil error if the measurement could not be marked as done
+	Done(msmt *DatabaseMeasurement) error
+
+	// UploadFailed writes the error string for the upload failure to the measurement
+	//
+	// Arguments:
+	//
+	// - msmt is the database measurement to update
+	//
+	// - failure is the error string to use
+	//
+	// Returns a non-nil error if the measurement update failed
+	UploadFailed(msmt *DatabaseMeasurement, failure string) error
+
+	// UploadSucceeded writes the error string for the upload failure to the measurement
+	//
+	// Arguments:
+	//
+	// - msmt is the database measurement to update
+	//
+	// Returns a non-nil error is measurement update failed
+	UploadSucceeded(msmt *DatabaseMeasurement) error
+
+	// Failed writes the error string to the measurement
+	//
+	// Arguments:
+	//
+	// - msmt is the database measurement to update
+	//
+	// - failure is the error string to use
+	//
+	// Returns a non-nil error if the measurement update failed
+	Failed(msmt *DatabaseMeasurement, failure string) error
+
+	// Close closes the database session
+	Close() error
```

Comment:
The same argument used above for `Session` applies to `Close`. Users of a database don't need to care about closing it. I think it's the piece of code that OWNS the database (i.e., the `ooni.Probe` struct) that should eventually be responsible of closing the database. All the rest of the codebase does not need to worry about closing the database and actually it should not even be capable of doing so, because it's outside its remit.